### ISSUE TITLE
Double facebook

### ISSUE THREAD

**Issue by mapkyca** [idno/Known#755]
As seen here: https://mapkyca.com/2015/that-moment-when-someone-asks-if-youre-familiar-with-a

I'm getting mentions posted twice, but in different formats - first just the mention, and then a "mentioned in".

Interestingly, these duplicates don't seem to count on the star or comment count.

Possibly a bridgy issue, but any thoughts?

**Comment by benwerd** [comment]
I've noticed it too. Not sure what's happening there, but double-posting does _seem_ to indicate that they're being sent twice.

**Comment by mapkyca** [comment]
Particularly as the content is different, and they get detected as different types - first is seen as a like, and the second as a comment of "liked this" for example

**Comment by mapkyca** [comment]
Seems to sugges two different pings with two different permalimks

**Comment by mapkyca** [comment]
Also, I've been getting many "new" pings for old stuff, or at least I'm getting emails, so I'm guessing something might be up with bridgy....

**Comment by mapkyca** [comment]
<a href="" class="u-in-reply-to"></a><a href="https://github.com/idno/Known/issues/755" class="u-in-reply-to"></a>Looking at the logs then, it does look like bridgy is pinging me multiple times for old posts - so that's either bridgy, or possibly facebook's API that's gone a little haywire.

/cc @snarfed ... thoughts? :)

**Comment by snarfed** [comment]
ugh, yeah, we've been discussing this too, in snarfed/bridgy#363 and earlier in snarfed/bridgy#346. snarfed/bridgy#350 may also be related. we *think* it's the FB API being weird, but we don't know yet. sorry for the dupes.

**Comment by mapkyca** [comment]
<a href="" class="u-in-reply-to"></a><a href="https://github.com/idno/Known/issues/755#issuecomment-74298780" class="u-in-reply-to"></a>No worries! Let me know if I can help at all (logs or whatever) :)

**Comment by mapkyca** [comment]
<a href="" class="u-in-reply-to"></a><a href="https://github.com/idno/Known/issues/755" class="u-in-reply-to"></a>Just an update, I'm now getting pinged for one specific message about once every half hour. 

Weird.

**Comment by snarfed** [comment]
@mapkyca sounds like retries. is known maybe returning 4xx or 5xx to the webmention but still emailing you? i think that's happened to known user(s) before.

**Comment by benwerd** [comment]
@mapkyca @snarfed Aha! That's interesting. Will check and see if Known is unnecessarily throwing an exception or something that's causing this behavior.

**Comment by mapkyca** [comment]
<a href="" class="u-in-reply-to"></a><a href="https://github.com/idno/Known/issues/755#issuecomment-74554269" class="u-in-reply-to"></a>According to my access.log, the bridgy posts have been received, and HTTP 202 returned. This includes our repeated value...

**Comment by snarfed** [comment]
aha, i see it. source https://brid-gy.appspot.com/comment/facebook/100001501388539/834537379939634/834537379939634_834791599914212 , target https://mapkyca.com/2015/my-bread-has-a-willy .

**Comment by snarfed** [comment]
response entity looks good and is marked complete. not sure why the propagate tasks don't see that and short circuit out.

https://appengine.google.com/datastore/edit?app_id=s~brid-gy&version_id=2.372689861983840287&key=aglzfmJyaWQtZ3lyQwsSCFJlc3BvbnNlIjV0YWc6ZmFjZWJvb2suY29tLDIwMTM6ODM0NTM3Mzc5OTM5NjM0XzgzNDc5MTU5OTkxNDIxMgw

(hijacking this issue to track progress, hope you all don't mind!)

**Comment by snarfed** [comment]
[example propagate log](https://www.brid.gy/log?start_time=1421967114&key=aglzfmJyaWQtZ3lyQwsSCFJlc3BvbnNlIjV0YWc6ZmFjZWJvb2suY29tLDIwMTM6ODM0NTM3Mzc5OTM5NjM0XzgzNDc5MTU5OTkxNDIxMgw):

```
 0.1.0.2 - - [22/Jan/2015:14:53:01 -0800] "POST /_ah/queue/propagate HTTP/1.1" 200 84 "http://brid-gy.appspot.com/_ah/queue/poll" "AppEngine-Google; (+http://code.google.com/appengine)"

Params: UnicodeMultiDict([(u'response_key', u'aglzfmJy...')])
Source: Marcus Povey (Facebook) 100001501388539, http://brid-gy.appspot.com/facebook/100001501388539
Created by this poll: http://brid-gy.appspot.com/log?start_time=1421967114&key=aglzfmJyaWQtZ3lyIQsSDEZhY2Vib29rUGFnZSIPMTAwMDAxNTAxMzg4NTM5DA
Starting Response comment tag:facebook.com,2013:834537379939634_834791599914212 https://www.facebook.com/834537379939634?comment_id=834791599914212
Webmention from https://brid-gy.appspot.com/comment/facebook/100001501388539/834537379939634/834537379939634_834791599914212 to https://mapkyca.com/2015/my-bread-has-a-willy
Sending...
Starting new HTTPS connection (1): mapkyca.com
"GET /2015/my-bread-has-a-willy HTTP/1.1" 200 None
Starting new HTTPS connection (1): mapkyca.com
"POST /webmention/ HTTP/1.1" 202 None
Sent! {'body': u'', 'request': u'POST https://mapkyca.com/webmention/ (with source=https://brid-gy.appspot.com/comment/facebook/100001501388539/834537379939634/834537379939634_834791599914212, target=https://mapkyca.com/2015/my-bread-has-a-willy)', 'http_status': 202}
Setting last_webmention_sent
```

**Comment by snarfed** [comment]
aha! we think the comment is changing every time, so we re-propagated it. bad diff logic. [example poll log](https://www.brid.gy/log?start_time=1424119790&key=aglzfmJyaWQtZ3lyIQsSDEZhY2Vib29rUGFnZSIPMTAwMDAxNTAxMzg4NTM5DA):

```
 0.1.0.2 - - [16/Feb/2015:12:49:55 -0800] "POST /_ah/queue/poll HTTP/1.1" 200 84 "http://www.brid.gy/_ah/queue/poll" "AppEngine-Google; (+http://code.google.com/appengine)"

Params: UnicodeMultiDict([(u'last_polled', u'2015-02-16-20-39-50'), (u'source_key', u'aglzfmJy...')])
Source: Marcus Povey (Facebook) 100001501388539, http://www.brid.gy/facebook/100001501388539
Last poll: http://www.brid.gy/log?start_time=1424119190&key=aglzfmJyaWQtZ3lyIQsSDEZhY2Vib29rUGFnZSIPMTAwMDAxNTAxMzg4NTM5DA
Using ETag "c7d5e603b806d223914814fe2021892e8c3d3e8c", last activity id 10205169725700870
...
https://www.facebook.com/100001501388539/posts/834537379939634 has 1 original post URL(s): https://mapkyca.com/2015/my-bread-has-a-willy
Response changed! Re-propagating. Original: Response(key=Key('Response', 'tag:facebook.com,2013:834537379939634_834791599914212'), activities_json=[u'{"url": "https://www.facebook.com/100001501388539/posts/834537379939634", "object": {...
Added propagate task: 7280927365356424374
```

**Comment by snarfed** [comment]
looks like it's person mention tags that are changing, either full name => first name or vice versa. e.g.

```
activity[tags]
[{u'displayName': u'Gina Rossman Barrett', u'url': u'https://www.facebook.com/100000224384191', u'length': 4, u'startIndex': 0, u'id': u'tag:facebook.com,2013:100000224384191', u'objectType': u'person'}]
=>
[{'displayName': u'Gina', 'url': u'https://www.facebook.com/100000224384191', 'startIndex': 0, 'length': 4, 'id': u'tag:facebook.com,2013:100000224384191', 'objectType': 'person'}]
```

**Comment by snarfed** [comment]
hopefully that fixes this. tentatively closing. feel free to reopen if it keeps happening.

btw @mapkyca, for the comment that kept re-pinging you, i think the source url and uid and content were always the same, which the webmention spec supports, and why i expect known de-duped them. i vaguely remember @benwerd saying he should suppress the email notification on dupe wms. that might be something to prioritize.

**Comment by mapkyca** [comment]
<a href="" class="u-in-reply-to"></a><a href="https://github.com/idno/Known/issues/755#issuecomment-74618324" class="u-in-reply-to"></a>Looks like it's working, I've not been pinged for that message since ~8 this morning. Thanks!

**Comment by benwerd** [comment]
Now this issue is done and dusted (awesome work!), I feel I'm able to respond with what I've been itching to say all thread:

_What does it mean???_

**Comment by snarfed** [comment]
lol at @benwerd.

tl;dr:
1. fixed a bridgy bug.
2. facebook (api) hates us so we can't have nice things.
3. sounds like known sends notifs for all successfull wms right now...? you might want to consider suppressing the notifs if they're pure dupes, ie they don't update or delete the existing copy. (you might also want to customize the notifs for updates and deletes, but you may do that already.)

**Comment by benwerd** [comment]
Ha, I was ham-fistedly making a double rainbow reference rather than actually wondering what was going on ;)

Known no longer sends double notifications. It's in code right now but not the release. The next release should also have that awesome bridgy bridge.

**Comment by mapkyca** [comment]
Ok, looks like we're almost there - I'm not getting the madness of constantly being pinged about one message, however it looks like I'm still being pinged twice for posts on Facebook.

So, comments, I seem to get two identical emails for. For likes, I get an email saying that a post has been liked, and another comment email containing the text "liked this", which appears as an additional comment on the post.

Looking at the output from Bridgy it doesn't *look* like I'm getting pinged twice, so I wonder whether this is a logic fault at the known side, perhaps a parsing problem?

**Comment by snarfed** [comment]
ah, got it. it's probably because the markup has two links to the original post, one as u-mention and one as u-like-of. e.g.:

```html
  <div class="e-content">  
  likes this.
<a class="u-mention" href="https://kylewm.com/10101859667632951/02/it-s-interesting-that-the-concept-of-a-one-hit-wonder"></a>
  </div>

<a class="u-like u-like-of" href="https://www.facebook.com/12802152/posts/10101214863845039"></a>
<a class="u-like u-like-of" href="https://kylewm.com/2015/02/it-s-interesting-that-the-concept-of-a-one-hit-wonder"></a>
```

**Comment by mapkyca** [comment]
Closing this, pretty sure it's fixed.
